### PR TITLE
71622 improve field validators

### PR DESCRIPTION
**Pull request by janid1967**

### CODE REVIEW

File: src/Equinor.Procosys.Preservation.Command/RequirementCommands/RecordValues/RecordValuesCommandValidator.cs [hunk 1]
```diff
@@ -32,13 +31,13 @@ public class RecordValuesCommandValidator : AbstractValidator<RecordValuesComman
             When(command => command.FieldValues.Any(), () =>
             {
                 RuleForEach(command => command.FieldValues)
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/RequirementCommands/RecordValues/RecordValuesCommandValidator.cs]: replace command with _ where command not used

File: src/tests/Equinor.Procosys.Preservation.Command.Tests/JourneyCommands/CreateStep/CreateStepCommandValidatorTests.cs [hunk 2]
```diff
@@ -115,33 +116,33 @@ public void Validate_ShouldFail_WhenResponsibleIsVoided()
         }
 
         [TestMethod]
-        public void Validate_ShouldFailWith3Errors_WhenErrorsInAllRules()
+        public void Validate_ShouldFailWith1Error_WhenErrorsInAllRules()
         {
```
Comment (janid1967) [inline src/tests/Equinor.Procosys.Preservation.Command.Tests/JourneyCommands/CreateStep/CreateStepCommandValidatorTests.cs]: rename to ..When3Errors

File: src/tests/Equinor.Procosys.Preservation.Command.Tests/TagCommands/CreateTag/CreateTagCommandValidatorTests.cs [hunk 3]
```diff
@@ -217,8 +217,8 @@ public void Validate_ShouldFail_WhenRequirementsNotUnique()
         [TestMethod]
         public void Validate_ShouldFailWith1Error_When2ErrorsWithinSameRule()
```
Comment (janid1967) [inline src/tests/Equinor.Procosys.Preservation.Command.Tests/TagCommands/CreateTag/CreateTagCommandValidatorTests.cs]: review other tests ala ShouldFailWith1Error and use this naming pattern

File: src/tests/Equinor.Procosys.Preservation.Command.Tests/Validators/ModeValidatorTests.cs [hunk 4]
```diff
@@ -1,118 +1,134 @@
...
-        private const string ModeVoidedTitle = "ModeVoided";
-        private const int ModeIdNonVoided = 1;
-        private const int ModeIdVoided = 2;
-        private ModeValidator _dut;
-
-        [TestInitialize]
-        public void Setup()
+        private const string ModeTitle = "TestMode";
+        private int ModeId;
+
```
Comment (janid1967) [inline src/tests/Equinor.Procosys.Preservation.Command.Tests/Validators/ModeValidatorTests.cs]: rename to _modeId

File: src/tests/Equinor.Procosys.Preservation.Command.Tests/Validators/RequirementDefinitionValidatorTests.cs [hunk 5]
```diff
@@ -1,66 +1,85 @@
...
-    public class RequirementDefinitionValidatorTests
+    public class RequirementDefinitionValidatorTests : ReadOnlyTestsBase
     {
-        private const int ReqDefIdNonVoided = 1;
-        private const int ReqDefIdVoided = 2;
-        private RequirementDefinitionValidator _dut;
-
-        [TestInitialize]
-        public void Setup()
+        private int ReqDefId;
```
Comment (janid1967) [inline src/tests/Equinor.Procosys.Preservation.Command.Tests/Validators/RequirementDefinitionValidatorTests.cs]: rename to _reqDefId

File: src/tests/Equinor.Procosys.Preservation.Command.Tests/Validators/ResponsibleValidatorTests.cs [hunk 6]
```diff
@@ -1,67 +1,85 @@
...
-    public class ResponsibleValidatorTests
+    public class ResponsibleValidatorTests : ReadOnlyTestsBase
     {
-        private const int ResponsibleIdNonVoided = 1;
-        private const int ResponsibleIdVoided = 2;
-        private ResponsibleValidator _dut;
-
-        [TestInitialize]
-        public void Setup()
+        private int ResponsibleId;
```
Comment (janid1967) [inline src/tests/Equinor.Procosys.Preservation.Command.Tests/Validators/ResponsibleValidatorTests.cs]: rename to _responsibleId

File: src/tests/Equinor.Procosys.Preservation.Command.Tests/Validators/StepValidatorTests.cs [hunk 7]
```diff
@@ -1,76 +1,86 @@
...
+    public class StepValidatorTests : ReadOnlyTestsBase
     {
-        private const string TestPlant = "PlantA";
-        private const int StepIdNonVoided = 1;
-        private const int StepIdVoided = 2;
-        private StepValidator _dut;
-
-        [TestInitialize]
-        public void Setup()
+        private int StepId;
```
Comment (janid1967) [inline src/tests/Equinor.Procosys.Preservation.Command.Tests/Validators/StepValidatorTests.cs]: _stepId

File: src/Equinor.Procosys.Preservation.Command/Validators/FieldValidators/FieldValidator.cs [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        private readonly IReadOnlyContext _context;
+
+        public FieldValidator(IReadOnlyContext context) => _context = context;
+
+        public async Task<bool> ExistsAsync(int fieldId, CancellationToken token)
+        {
+            var count = await (from f in _context.QuerySet<Field>()
+                where f.Id == fieldId
+                select f).CountAsync(token);
+            return count > 0;
```
Comment (hbjorgo) [inline src/Equinor.Procosys.Preservation.Command/Validators/FieldValidators/FieldValidator.cs]: .Any() is probably much faster than Count(). Count must fin all elements, Any() must only find one. This should be fixed in all these scenarios.

File: src/Equinor.Procosys.Preservation.Command/Validators/JourneyValidators/JourneyValidator.cs [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        private readonly IReadOnlyContext _context;
+
+        public JourneyValidator(IReadOnlyContext context) => _context = context;
+
+        public async Task<bool> ExistsAsync(int journeyId, CancellationToken token)
+        {
+            var count = await (from j in _context.QuerySet<Journey>()
+                where j.Id == journeyId
+                select j).CountAsync(token);
+            return count > 0;
```
Comment (hbjorgo) [inline src/Equinor.Procosys.Preservation.Command/Validators/JourneyValidators/JourneyValidator.cs]: Ref. earlier comment.

File: src/Equinor.Procosys.Preservation.Command/Validators/JourneyValidators/JourneyValidator.cs [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+        }
+
+        public async Task<bool> ExistsAsync(string title, CancellationToken token)
+        {
+            var count = await (from j in _context.QuerySet<Journey>()
+                where j.Title == title
+                select j).CountAsync(token);
+            return count > 0;
```
Comment (hbjorgo) [inline src/Equinor.Procosys.Preservation.Command/Validators/JourneyValidators/JourneyValidator.cs]: Ref. earlier comment.

File: src/Equinor.Procosys.Preservation.Command/Validators/ModeValidators/ModeValidator.cs [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+        private readonly IReadOnlyContext _context;
+
+        public ModeValidator(IReadOnlyContext context) => _context = context;
+
+        public async Task<bool> ExistsAsync(int modeId, CancellationToken token)
+        {
+            var count = await (from m in _context.QuerySet<Mode>()
+                where m.Id == modeId
+                select m).CountAsync(token);
+            return count > 0;
```
Comment (hbjorgo) [inline src/Equinor.Procosys.Preservation.Command/Validators/ModeValidators/ModeValidator.cs]: Ref. earlier comment.

File: src/Equinor.Procosys.Preservation.Command/Validators/ModeValidators/ModeValidator.cs [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+        }
+        
+        public async Task<bool> ExistsAsync(string title, CancellationToken token)
+        {
+            var count = await (from m in _context.QuerySet<Mode>()
+                where m.Title == title
+                select m).CountAsync(token);
+            return count > 0;
```
Comment (hbjorgo) [inline src/Equinor.Procosys.Preservation.Command/Validators/ModeValidators/ModeValidator.cs]: Ref. earlier comment.

File: src/Equinor.Procosys.Preservation.Command/Validators/ModeValidators/ModeValidator.cs [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+                select m).FirstOrDefaultAsync(token);
+            return mode != null && mode.IsVoided;
+        }
+
+        public async Task<bool> IsUsedInStepAsync(int modeId, CancellationToken token)
+        {
+            var count = await (from s in _context.QuerySet<Domain.AggregateModels.JourneyAggregate.Step>()
+                where s.ModeId == modeId
+                select s).CountAsync(token);
+            return count > 0;
```
Comment (hbjorgo) [inline src/Equinor.Procosys.Preservation.Command/Validators/ModeValidators/ModeValidator.cs]: Ref. earlier comment.

File: src/Equinor.Procosys.Preservation.Command/Validators/RequirementDefinitionValidators/RequirementDefinitionValidator.cs [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+        private readonly IReadOnlyContext _context;
+
+        public RequirementDefinitionValidator(IReadOnlyContext context) => _context = context;
+
+        public async Task<bool> ExistsAsync(int requirementDefinitionId, CancellationToken token)
+        { 
+            var count = await (from rd in _context.QuerySet<RequirementDefinition>()
+                where rd.Id == requirementDefinitionId
+                select rd).CountAsync(token);
+            return count > 0;
```
Comment (hbjorgo) [inline src/Equinor.Procosys.Preservation.Command/Validators/RequirementDefinitionValidators/RequirementDefinitionValidator.cs]: Ref. earlier comment.

File: src/Equinor.Procosys.Preservation.Command/Validators/ResponsibleValidators/ResponsibleValidator.cs [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+        private readonly IReadOnlyContext _context;
+
+        public ResponsibleValidator(IReadOnlyContext context) => _context = context;
+
+        public async Task<bool> ExistsAsync(int responsibleId, CancellationToken token)
+        {
+            var count = await (from r in _context.QuerySet<Responsible>()
+                where r.Id == responsibleId
+                select r).CountAsync(token);
+            return count > 0;
```
Comment (hbjorgo) [inline src/Equinor.Procosys.Preservation.Command/Validators/ResponsibleValidators/ResponsibleValidator.cs]: Ref. earlier comment.

File: src/Equinor.Procosys.Preservation.Command/Validators/StepValidators/StepValidator.cs [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+        private readonly IReadOnlyContext _context;
+
+        public StepValidator(IReadOnlyContext context) => _context = context;
+
+        public async Task<bool> ExistsAsync(int stepId, CancellationToken token)
+        {
+            var count = await (from s in _context.QuerySet<Step>()
+                where s.Id == stepId
+                select s).CountAsync(token);
+            return count > 0;
```
Comment (hbjorgo) [inline src/Equinor.Procosys.Preservation.Command/Validators/StepValidators/StepValidator.cs]: Ref. earlier comment.

**Pull request merged**